pkg/ccl/backupccl/restore_planning.go

```diff
@@ -158,6 +161,242 @@ func synthesizePGTempSchema(
+// sequence value that is greater than the IDs of all the passed in descriptors.
+func bumpDescIDSequenceKey(
+	ctx context.Context,
+	p sql.PlanHookState,
+	databasesByID map[descpb.ID]*dbdesc.Mutable,
+	schemasByID map[descpb.ID]*schemadesc.Mutable,
+	tablesByID map[descpb.ID]*tabledesc.Mutable,
+	typesByID map[descpb.ID]*typedesc.Mutable,
+) error {
+	var maxDescIDInBackup int64
+	// Include the table descriptors when calculating the max ID.
+	for _, table := range tablesByID {
+		if int64(table.ID) > maxDescIDInBackup {
+			maxDescIDInBackup = int64(table.ID)
+		}
+	}
+
+	// Include the database descriptors when calculating the max ID.
+	for _, database := range databasesByID {
+		if int64(database.ID) > maxDescIDInBackup {
+			maxDescIDInBackup = int64(database.ID)
+		}
+	}
+
+	// Include the type descriptors when calculating the max ID.
+	for _, typ := range typesByID {
+		if int64(typ.ID) > maxDescIDInBackup {
+			maxDescIDInBackup = int64(typ.ID)
+		}
+	}
+
+	// Include the schema descriptors when calculating the max ID.
+	for _, sc := range schemasByID {
+		if int64(sc.ID) > maxDescIDInBackup {
+			maxDescIDInBackup = int64(sc.ID)
+		}
+	}
+
+	return p.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		v, err := txn.Get(ctx, p.ExecCfg().Codec.DescIDSequenceKey())
+		if err != nil {
+			return err
+		}
+		newValue := maxDescIDInBackup + 1
+		if newValue <= v.ValueInt() {
+			// This case may happen when restoring backups from older versions.
+			newValue = v.ValueInt() + 1
+		}
+		b := txn.NewBatch()
+		// N.B. This key is usually mutated using the Inc command. That
+		// command warns that if the key was every Put directly, Inc will
+		// return an error. This is only to ensure that the type of the key
+		// doesn't change. Here we just need to be very careful that we only
+		// write int64 values.
+		// The generator's value should be set to the value of the next ID
+		// to generate.
+		b.Put(p.ExecCfg().Codec.DescIDSequenceKey(), newValue)
+		return txn.Run(ctx, b)
+	})
+}
+
+// findConflictingSystemTables collects all the tables in the `system` database
+// and checks if their ID collides with any of the passed in descriptor IDs. The
+// method returns a mapping from the colliding ID to the colliding system table
+// descriptor.
+func findConflictingSystemTables(
+	ctx context.Context,
+	p sql.PlanHookState,
+	databasesByID map[descpb.ID]*dbdesc.Mutable,
+	schemasByID map[descpb.ID]*schemadesc.Mutable,
+	tablesByID map[descpb.ID]*tabledesc.Mutable,
+	typesByID map[descpb.ID]*typedesc.Mutable,
+) (map[descpb.ID]catalog.TableDescriptor, error) {
+	systemTablesWithConflictingIDs := make(map[descpb.ID]catalog.TableDescriptor)
+	err := sql.DescsTxn(ctx, p.ExecCfg(), func(ctx context.Context, txn *kv.Txn, col *descs.Collection) error {
+		systemTableDescs, err := col.GetAllTableDescriptorsInDatabase(ctx, txn, systemschema.SystemDB.GetID())
```

Comment:
I still sorta wonder if we should just be looking for _any_ desc that has an ID in our maps, rather than just system descs, but I'll grant that it is the same in practice for 22.1 and we expect to not need this for 22.2 anyway, so maybe moot.

Ditto on getting just _table_ descriptors vs any descriptor that has a matching ID, but again, maybe moot since right now they're all tables anyway.

Reply:
I have a TODO after the call to `moveConflictingSystemTables` to do a sanity check that none of the descriptors in the `system.descriptors` table conflict with our backed up descriptors. In which case we should just fail the restore.

But yeah if we're getting rid of all this logic imminently then I'd rather not deal with moving types, schemas, databases at the moment. Does that sound reasonable?

Reply:
yeah, I think this is fine to punt.